mesonbuild/interpreter/interpreter.py

```diff
@@ -1019,16 +1029,29 @@ def set_backend(self):
         options = {k: v for k, v in self.environment.options.items() if k.is_backend()}
         self.coredata.set_options(options)
 
-    @permittedKwargs({'version', 'meson_version', 'default_options', 'license', 'subproject_dir'})
     @typed_pos_args('project', str, varargs=str)
-    def func_project(self, node: mparser.FunctionNode, args: T.Tuple[str, T.List[str]], kwargs: 'TYPE_kwargs') -> None:
+    @typed_kwargs(
+        'project',
+        DEFAULT_OPTIONS,
+        KwargInfo('meson_version', (str, NoneType)),
+        KwargInfo(
+            'version',
+            (str, mesonlib.File, NoneType, list),
+            default='undefined',
+            validator=_project_version_validator,
+            convertor=lambda x: x[0] if isinstance(x, list) else x,
```

Comment:
What all is this even doing? We don't document this accepts lists at all.

> when passed as array must have a length of 1

Why does this warning message exist like that?

Is there some perfectly dreadful reason we need to keep supporting single-element list literals here for backwards compat, even though we never allowed it and anyone doing it was completely wrong?



```python
        version = kwargs.get('version', 'undefined')
        if isinstance(version, list):
            if len(version) != 1:
                raise InvalidCode('Version argument is an array with more than one entry.')
            version = version[0]
```

This smells to me like badly designed error checking, because what we actually want is: 3f0a0c1582f55e0da950a7a321d25ffcf240ec6f

We can't directly accept a mesonlib.File seemingly because files() returns a list of them (but people can technically do `files('VERSION')[0]`, it just looks odd, so maybe we *should* keep accepting that type), but we should be checking, if it is a list, that it has exactly one element and that element is a File object, *not* a string.

Reply:
I tried to get rid of it, but found some case where we really did need a list, but I can't remember now what it was

Reply:
The original "support" for lists was only added in the same commit that added files() support -- again, files() returns an `array[file]`, so what we actually want to accept here is `str | array[file]`.

Reply:
That's it. And that's a lie of work just avoid taking an array of string

Reply:
Gah, typing in a moving car is a pain

Reply:
It's not a lot of work, it's just removing `str` from the validator.

Reply:
By work, it seems like saying `project(version : [files('foo')])` Okay! but `project(version : ['0.1.2'])`, nope! is a lot of work, somewhat confusing, and just not that useful? :shrug:

Reply:
Actually `[files('foo')]` becomes `array[array[file]]` and was illegal either way. :D

Reply:
But it get's flattened to `[]file`, right?